### DISCUSSION TITLE
Usability for Linux Mobile

### DISCUSSION THREAD

**Discussion by tobtobxx** [emilk/egui#788] [Q&A] [answered]
So, I have a phone that runs Linux (not Android). (It's a [Pinephone](https://www.pine64.org/pinephone/).)

I'd like to develop applications for it, and stumbled over egui. There's a problem though: egui_glim doesn't seem to register touches.

Do you have any recommendations for which backend to try? And can I use `epi` with another backend?

**Comment by emilk** [comment]
In `egui-winit` (used by `egui_glium`) does listen for touch events, but perhaps `winit` doesn't support android touches yet. I would raise (or find) an issue on https://github.com/rust-windowing/winit/issues

You could try `egui_web` instead (perhaps using a [progressive web app](https://github.com/emilk/egui_template/pull/15)?)

**Reply by tobtobxx** [reply]
Hmm..., I hijacked [`egui_winit::State::on_event`](https://github.com/emilk/egui/blob/fe2094af14e6709b0ee178aadbf92384438220a5/egui-winit/src/lib.rs#L187) to print the events it receives, and it does receive the touch events.

I also plugged in a mouse and clicking with a mouse worked fine, just via touch it doesn't work.

Btw, it is not an Android phone, it runs mainline Linux and you install distros on that thing. So it probably gets recognized by almost everything as Dekstop with a weird form factor. Due to low specs, PWAs (and general web technologies) are rather unusable.

Could you pinpoint me to where normally the conversion from touch to pointer input happens (not for web, for desktop)? I'll try to troubleshoot from there...

**Reply by emilk** [reply]
Oh, you wrote _not_ Android 😅

Touch was introduced in https://github.com/emilk/egui/pull/306, but I have only ever tested the `egui_web` touch input, so I suspect the problem is in `egui-winit`.

egui expects the integrations to pass in the touch events but also pointer events when there is touch input. `egui_web` does so in its touch handler, but I see that `egui-winit` does not.

When you touch, do you get both `WindowEvent::Touch` and `WindowEvent::MouseInput`? If not, that's the problem, and we need to also add some pointer events in the `on_touch` handler in `egui-winit` (`egui::event::PointerMoved, PointerButton, PointerGone`).

**Reply by tobtobxx** [reply]
I indeed only get the `WindowEvent::Touch` events and no `WindowEvent::MouseInput` at location I linked above (btw, [this](https://github.com/TobTobXX/egui/commit/d93b418520c088ac01031515167b2683bd178f9d) is the specific patch I made to debug this).

For the logic: I think it would make the most sense to just send a `MouseInput` with `state: Pressed` on touch down and with `state: Released` on touch up. Is it also necessary to convert the moves into `CursorMoved` events?

Also: What should happen if you touch at two places? Is it allowed to have two pointers (although, if it is allowed on egui_web, then, yeah probably).

**Reply by tobtobxx** [reply]
Yeah, I wrote a primitive PoC, and that was indeed the issue: https://github.com/TobTobXX/egui/commit/6073196ddd979514bb50059aa2ee914aaec7808f

With this change, touching the button makes it being clicked (and remain that way, bc there's no release).

I'm gonna try to write a proper patch and submit a PR...

**Comment by tobtobxx** [comment, marked as answer]
Is being fixed in this PR: #791

**Comment by Barugon** [comment]
> So, I have a phone that runs Linux (not Android). (It's a [Pinephone](https://www.pine64.org/pinephone/).)

How do you get this to work at all? I compile this basic app on the PinePhone Pro (Mobian) and run it, I see the window decorations appear and then the phone restarts.

```Rust
use eframe::egui;

#[derive(Default)]
struct App;

impl eframe::App for App {
    fn update(&mut self, _ctx: &egui::Context, _frame: &mut eframe::Frame) {}
}

fn main() {
    eframe::run_native(
        "Test",
        eframe::NativeOptions::default(),
        Box::new(|_cc| Box::new(App::default())),
    );
}
```

**Reply by Barugon** [reply]
I tried the wgpu feature and I got the same issue.

**Reply by tobtobxx** [reply]
Wait... the _phone restarts_?! As in the DE, OS or Kernel crashes? Wow... I used Phosh on PostmarketOS, so there's that. I'm planning to work on it again and then I can give you more infos on the setup, but that's gonna take a while (~2-4mo). Until then, I can't help you much unfortunately. (If you need a rubber duck u can still try to hit me up tho.)

**Reply by Barugon** [reply]
> Wait... the _phone restarts_?!

Yup, total restart. I don't even know how to debug this. I'm working on a VFR flight planner and I really wanted the PinePhone to be a primary target.

**Reply by tobtobxx** [reply]
> I don't even know how to debug this.

I'd start with sshing in and then watching `dmesg -w` or `journalctl -f`. If the phone restarts, then the rest of the system should take notice. (Btw, syslogs are persistent across reboots/crashes, so you can also pull up `journalctl -b1` for the last uptime period and analyze the end of the log.)

**Reply by Barugon** [reply]
Thanks for the info.

**Comment by Barugon** [comment]
I got it somewhat working on Manjaro. Here's what it should look like (simulated proportions on desktop):

![screenshot](https://user-images.githubusercontent.com/16503728/188284324-c9abb2b2-af1a-4297-b7ce-542c5e4a8276.png)

And here's what it actually looks like on the device:

![screenshot_device](https://user-images.githubusercontent.com/16503728/188284422-7014a7c5-cf96-4b86-ae47-939c8cee9ce5.png)

If I run my finger around  in the window then it segfaults.  :-(

**Reply by Barugon** [reply]
Here's what I get in a debugger:

![screenshot_sigsegv](https://user-images.githubusercontent.com/16503728/188515868-632f7491-ba8c-4a0f-b21c-fd0ea4763318.png)

**Reply by Barugon** [reply]
Regular PinePhone works, so this is probably a driver issue on the Pro.

**Reply by tobtobxx** [reply]
> Here's what I get in a debugger:

Did you run it in release mode? Try to crash it again in debug mode, as that might give more useful information on crash.

Also, for a primer on how to get useful informations out of a segfault, check this out: https://jvns.ca/blog/2018/04/28/debugging-a-segfault-on-linux/

Can't help, as I don't own a PPP, sorry.

**Reply by Barugon** [reply]
That was debug mode.

**Reply by Barugon** [reply]
Since it works on the non-Pro PinePhone, I'm going to try to figure out why pinch zoom doesn't work.
